src/Engine/UnitActionType.java

```diff
@@ -25,9 +25,35 @@
   public static final UnitActionType[] COMBAT_VEHICLE_ACTIONS = { ATTACK,           WAIT, LOAD, JOIN };
   public static final UnitActionType[] TRANSPORT_ACTIONS =      { UNLOAD,           WAIT, LOAD, JOIN };
   public static final UnitActionType[] APC_ACTIONS =            { UNLOAD, RESUPPLY, WAIT, LOAD, JOIN };
-  
 
-  public static class Attack implements UnitActionType
+  /**
+   * Base type to standardize indexing into HashMaps
+   */
+  abstract static class BaseUnitActionType implements UnitActionType
+  {
+    @Override
+    public int hashCode()
+    {
+      return name().hashCode();
+    }
+
+    @Override
+    public boolean equals(Object obj)
+    {
+      if( this == obj )
+        return true;
+      if( obj == null )
+        return false;
+      if( getClass() != obj.getClass() )
+        return false;
+      UnitActionType other = (UnitActionType) obj;
```

Comment:
Probably worth checking that `null != other` as well, in case the cast fails.

Reply:
Hmm.
We just checked that the classes are equal, and I'm casting to an interface that this class implements.
I can't see a case where that could happen.

Reply:
Ah. Right. I blame being awake at 5 am.